### DISCUSSION TITLE
Issues with eigenvalues and spatial distribution of 3D hydrogen atom

### DISCUSSION THREAD

**Discussion by Ionizing** [quantum-visualizations/qmsolve#9] [Q&A] [answered]
Hi, 
I just found this wonderful repo and it helped me a lot in understanding the spatial distribution of each level of the hydrogen atom.

 But when I ran the script `3D_hydrogen_atom.py`, I  found two issues with this program:
1. The eigenvalues of it returned seemed to be far from the known values: 
<img width="584" alt="image" src="https://user-images.githubusercontent.com/17506747/122521896-f1a04500-d047-11eb-8345-524fd1af85e4.png">

The ground state should be -13.6eV, but the script returned -8.26297717eV. Also, I found the grid is 30*30*30 with the actual length of each edge are also 30A*30A*30A. Should I make a finner grid to solve for a better result?

2. The axis unit is much larger than the known radial distribution. According to the radial distribution of hydrogen 1s state. There is a peak below the Bohr radius (0.53A). But the extent of of the result is much larger
<img width="529" alt="image" src="https://user-images.githubusercontent.com/17506747/122523233-632cc300-d049-11eb-97c0-f98622e76658.png">
However, if I change the `extent` parameter to 5A, the issue still remains:
<img width="546" alt="image" src="https://user-images.githubusercontent.com/17506747/122523795-ff56ca00-d049-11eb-9e79-bc567f9c9b77.png">

Though I came with such issues, I must say thank you for giving a feasible numeric approach to solve the 3D single particle schrodinger equation and giving fancy results to help students understanding the QM.

Looking forward to your reply!

**Comment by rafael-fuente** [comment, marked as answer]
The problem is the spatial discretization, as you said 30x30x30 is very small, especially considering that the Coulomb potential has a singularity. The problem is fixed using a smoother discretization, but If you try a discretization smoother than (50x50x50) you will soon find that a personal computer cannot handle it. 

But I already fixed the problem!

It happens that LOBPCG algorithm can deal with very large grids (200x200x200) and improve the accuracy of the solver considerably. I'm working on implementing it. I will update the repository with the implementation this weekend.

Using a 200x200x200 grid these are energies:

`[-13.52,  -3.384,  -1.502,  -0.826,   -0.531] eV`

while the analytical ones are:

`[13.605     ,  3.401  ,  1.512,  0.850,  0.544] eV`

This is the best you will be able to get with a 3D simulation. ( the computation last about 100 seconds on a good pc)

But the hydrogen atom has a clearly azimuthal symmetry and the problem can be reduced to 2D.
I'm currently working on an azimuthal solver too. This is what I get using a 700x700 grid:

`[-13.58,  -3.394,  -1.504,   -0.848,   -0.543] eV`

**Reply by Ionizing** [reply]
I tried 3D hydrogen atom system with 200*200*200 grid using 'lobpcg' method, but ended up with `[-1.79820075  1.89885966  8.35073314] eV`. Maybe there was something wrong with my input.

Looking forward to your update. And I'll appreciate it if the examples are updated too.

Thank you very much.

**Reply by rafael-fuente** [reply]
The current implemented lobpcg method is unstable for most configurations, or it requires a very large number of iterations to converge when the max iterations allowed are fixed to 20.  

I already fixed the problem using as an initial guess the 'eighs' eigenvectors computed with a small grid interpolated on a large one. If everything is fine, tomorrow I'll update the repository, including the examples.

**Reply by rafael-fuente** [reply]
I updated the repository with the new lobpcg implementation and updated the examples.  If you have GPU I highly recommend running the cupy lobpcg solver, because the speed boost is huge.

For using the cupy lobpcg solver use the argument `method ='lobpcg-cupy'`
Example:
`eigenstates = H.solve( max_states = 10, N0 = 30, method ='lobpcg-cupy')
`

**Reply by Ionizing** [reply]
Thanks for the update!

# The 3D system
I tried to reproduce the results of `3D_hydrogen_atom.py` using `lobpcg-cupy` and removed external electric field:

```
N = 150
[-12.72161905  -3.38207744  -3.38207744  -3.38207744  -3.29206952
  -1.49646744  -1.49646744  -1.49646744  -1.49391049  -1.49391049]

N = 200
[-13.04478545  -3.38147744  -3.38147744  -3.38147744  -3.32281329
  -1.49616019  -1.49616019  -1.48773828  -1.48773828  -1.48773828]

N = 220
[-13.12419074  -3.38184569  -3.38184569  -3.38184569  -3.32780468
  -1.49671495  -1.49671495  -1.48514553  -1.48514553  -1.48514553]
```

I've already use `r = np.where(r<0.000001, 0.000001, r)` to represent the singularity. It seems there is still something wrong with my input file.

# The 2D system
May I ask that if I should change `k_c` or something else in 2D system?

Here is what I wrote:

```
import numpy as np
from qmsolve import Hamiltonian, SingleParticle, init_visualization


#interaction potential
def coulomb_potential(particle):

    k_c = 14.39964547842567 # (e*e / (4 * np.pi * epsilon_0))  # measured in eV / Å

    r = np.sqrt((particle.x)**2 + (particle.y)**2)
    r = np.where(r < 0.0001, 0.0001, r)
    return - k_c/ r




H = Hamiltonian(particles = SingleParticle(),
                potential = coulomb_potential,
                spatial_ndim = 2, N = 700, extent = 30)


eigenstates = H.solve(max_states = 20)
print(eigenstates.energies)
```

Result:
```
[-47.43938949  -6.03309168  -6.03309168  -5.76634202  -2.17165797
  -2.17165797  -2.17099722  -2.17092577  -2.11324357  -1.10207235
  -1.10207235  -1.0960321   -1.07930107  -1.0744426   -1.0744426
  -1.04529177  -0.60149706  -0.59054392  -0.50565554  -0.50565554]
```

Seems not correct ....

**Reply by rafael-fuente** [reply]
Your 3D example is right. - 3.38 eV appears repeated three times because these levels are degenerate. They correspond to 2px, 2py, and 2pz orbitals. The analytical energy is -3.40 eV.

The four eigenstate is -3.33 eV. It corresponds to the orbital 2s. Ideally, it should be also degenerated with the 2p orbitals, but because of the finite size of the grid, it has a little more energy. However, you can notice that is still close to the real value.

The rest of the energies correspond to n = 3 orbitals, which are all degenerate.


The ground state has the worst precision (3.5% of error) You can improve it by making the extent of the size smaller to fit the small size of the orbital. For example, using `extent = 8` , by running the following script I obtain:
```
import numpy as np
from qmsolve import Hamiltonian, SingleParticle, init_visualization


#interaction potential
def coulomb_potential(particle):

	k_c = 14.39964547842567 # (e*e / (4 * np.pi * epsilon_0))  # measured in eV / Å

	r = np.sqrt((particle.x)**2 + (particle.y)**2 + (particle.z)**2)
	r = np.where(r < 0.00001, 0.00001, r)
	external_electric_field = 0.00000 #shows Stark effect

	return - k_c/ r + particle.z*external_electric_field


H = Hamiltonian(particles = SingleParticle(), 
				potential = coulomb_potential, 
				spatial_ndim = 3, N = 220, extent = 8)



eigenstates = H.solve( max_states = 2, N0 = 40, method ='lobpcg-cupy' , maxiter = 60)
print(eigenstates.energies)


```

`ground state = -13.46144372 eV
`

The error is now reduced to 1%.

About the 2D system you send me, it works right, but it doesn't represent the hydrogen atom. When you define potential in 2D it's assumed that potential doesn't vary the z coordinate. Your potential doesn't represent the Coulomb potential because it should be a function of z.

The way to simplify the hydrogen atom to a 2D is by separating the azimuthal part (cylindrical symmetry), so the potential only depends on z a p coordinates  (where `p = sqrt(x*x + y*y)`.  
This cannot be done with the current solver, but I'm going to add the feature in a near future.

**Reply by Ionizing** [reply]
> You can improve it by making the extent of the size smaller to fit the small size of the orbital.  For example, using `extent = 8`

I get to figure out that why it is hard to obtain the accurate result of the ground state of 3D_hydrogen_atom: it's hard to process the singularity of Coulomb potential. When we use the finner grid, the smaller `r` we can use. I checked my results again and found that except the ground state level, all the other `n>1` state level is much closer to the analytical result: `3.38207744 - (13.6 / 4) ~= -0.017eV`, `1.49646744 - (13.6 / 9) ~= -0.014eV`. I want to ask that if the spatial distribution of eigenstates is zoomed in the visual window because I saw that n=2,3 states can spread to 5 angstroms or more. And if I choose `extent = 8`, will it have an impact on the spatial distribution (e.g. the eigenstates are squeezed)?

**Reply by Ionizing** [reply]
I just re-implemented the solving part of 3D_hydrogen_atom in `julia`, see [here](https://gist.github.com/Ionizing/4a58e5d94b4a57b01fca9e13479f3ae9). A vast performance improvement was observed (It took only 32 secs to calculate a `50*50*50` system using `eigs` instead of `lobpcg` on a laptop). I am not sure if there are still some bugs in it, but the eigen values seemed to be reasonable.

**Reply by rafael-fuente** [reply]
Yes, singularity is a big problem. We need a large grid to process accurately it.  You may find useful this visualization of the radial distribution of hydrogen ground state, in which I compared the results of a simulation with extent = 16, and N = 200 with the analytical expression:

![comparision_analytical_numerical](https://user-images.githubusercontent.com/37953893/123012993-92359280-d3c3-11eb-803f-ff4d907cc70a.png)

As you can see it compares quite well with the analytical value. However, there is a little disagreement near the singularity. This doesn't happen for the other orbitals, because its wavefunction distribution is further from the singularity.

Here also a  3d density plot:
![snapshot](https://user-images.githubusercontent.com/37953893/123013486-9e6e1f80-d3c4-11eb-9039-5336730843d8.png)

> I want to ask that if the spatial distribution of eigenstates is zoomed in the visual window because I saw that n=2,3 states can spread to 5 angstroms or more. 

It's completely fine, they grow up pretty fast. Exactly, the mean growth of the radius of the hydrogen orbitals grows proportionally to the square of the quantum number n. Here an exact formula that illustrates this fact:

![radial-distribution-formula](https://user-images.githubusercontent.com/37953893/123013790-2e13ce00-d3c5-11eb-8d15-c3133d9dac42.png)

> And if I choose extent = 8, will it have an impact on the spatial distribution (e.g. the eigenstates are squeezed)?

Yes. You wouldn't be simulating a hydrogen atom, but a hydrogen atom on a box. The extent must be always larger than the size of the wave function.

> I just re-implemented the solving part of 3D_hydrogen_atom in `julia`, see [here](https://gist.github.com/Ionizing/4a58e5d94b4a57b01fca9e13479f3ae9). A vast performance improvement was observed (It took only 32 secs to calculate a `50*50*50` system using `eigs` instead of `lobpcg` on a laptop). I am not sure if there are still some bugs in it, but the eigen values seemed to be reasonable.

That's pretty nice! Definitely, a port of the solver to Julia may worth it.

**Reply by Ionizing** [reply]
Okay, it seems all of my problems are already solved. Thank you very much!

**Reply by Ionizing** [reply]
Could I ask one more question? How to construct the kinetic matrix in Hamiltonian when `Nx != Ny != Nz`? I read the code and found:
```python
def get_kinetic_matrix(self, H):  # Single particle kinetic matrix

    I = eye(H.N)
    T_ =  diags([-2., 1., 1.], [0,-1, 1] , shape=(H.N, H.N))*-k/(self.m*H.dx**2)
    if H.spatial_ndim ==1:
        T = T_

    elif H.spatial_ndim ==2:
        T =  kron(T_,I) + kron(I,T_)

    elif H.spatial_ndim ==3:
        T =  kron(T_, kron(I, I)) + kron(I, kron(T_, I)) + kron(I, kron(I, T_))

    return T
```
In my view, I should constuct not only the kinetic matrices for each axis: `Tx_, Ty_, Tz_`, but also the Identity matrices `Ix, Iy, Iz`. Then the total kinetic matrix should be `T = kron(Tx_, kron(Iy, Iz)) + kron(Ix, kron(Ty_, Iz)) + kron(Ix, kron(Iy, Tz_))`. Is there anything wrong with my derivation? Are there any papers or books to help understand the math?

**Reply by rafael-fuente** [reply]
> Could I ask one more question? 

Of course! :)

> In my view, I should constuct not only the kinetic matrices for each axis: Tx_, Ty_, Tz_, but also the Identity matrices Ix, Iy, Iz. Then the total kinetic matrix should be T = kron(Tx_, kron(Iy, Iz)) + kron(Ix, kron(Ty_, Iz)) + kron(Ix, kron(Iy, Tz_)). Is there anything wrong with my derivation? Are there any papers or books to help understand the math?

We didn't implement it, but different spatial discretization for each axis may be useful for some systems. Your guess is completely right. 
Here is some introduction to the math:

Discretizing the Hamiltonian:
 - [Discretization of 1D continuous Hamiltonian](https://wiki.physics.udel.edu/phys824/Discretization_of_1D_continuous_Hamiltonian)

Using 2D Grids for higher dimensional discrete operators:
 - [Kronecker sum of discrete Laplacians](https://en.wikipedia.org/wiki/Kronecker_sum_of_discrete_Laplacians)

**Reply by Ionizing** [reply]
Thanks for the reply! The supplementary links are very helpful!
